server/src/main/java/org/elasticsearch/common/xcontent/support/XContentMapValues.java

```diff
@@ -249,6 +250,81 @@ private static Object extractValue(String[] pathElements, int index, Object curr
         return null;
     }
 
+    /**
+     * Insert or replace the path's value with the provided new value. The map will be modified in-place.
+     * If the path does not exist in the map, it will be added as a flat top-level entry with the provided new value.
+     *
+     * @param path the value's path in the map.
+     * @param map the map to search and modify in-place.
+     * @param newValue the new value to assign to the path.
+     */
+    public static void insertValue(String path, Map<?, ?> map, Object newValue) {
+        String[] pathElements = path.split("\\.");
+        if (pathElements.length == 0) {
+            return;
+        }
+
+        List<SuffixMap> buffer = new ArrayList<>();
+        extractSuffixMaps(pathElements, 0, map, buffer);
+
+        if (buffer.isEmpty()) {
+            // No suffix maps were extracted, meaning the path was not found in the map. Add the new value using a flat path to handle if
+            // subobjects: false in downstream parsing.
+            @SuppressWarnings("unchecked")
+            Map<String, Object> castedMap = (Map<String, Object>) map;
+            castedMap.put(path, newValue);
+        } else if (buffer.size() == 1) {
+            SuffixMap suffixMap = buffer.getFirst();
+            suffixMap.map().put(suffixMap.suffix(), newValue);
+        } else {
+            throw new IllegalArgumentException(
+                "Path [" + path + "] matches " + buffer.size() + " values, it is ambiguous which value to replace"
+            );
+        }
+    }
+
+    private record SuffixMap(String suffix, Map<String, Object> map) {}
+
+    private static boolean extractSuffixMaps(String[] pathElements, int index, Object currentValue, List<SuffixMap> buffer) {
+        if (index == pathElements.length) {
+            return true;
+        } else if (currentValue instanceof List<?> valueList) {
+            for (Object o : valueList) {
+                extractSuffixMaps(pathElements, index, o, buffer);
+            }
+
+            return false;
+        } else if (currentValue instanceof Map<?, ?>) {
+            @SuppressWarnings("unchecked")
+            Map<String, Object> map = (Map<String, Object>) currentValue;
+
+            String key = pathElements[index];
+            while (index < pathElements.length) {
+                if (map.containsKey(key)) {
+                    if (extractSuffixMaps(pathElements, index + 1, map.get(key), buffer)) {
+                        buffer.add(new SuffixMap(key, map));
+                    }
+                }
+
+                if (++index < pathElements.length) {
+                    key += "." + pathElements[index];
+                }
+            }
+
```

Comment:
I wonder if we should still add the suffix map if the path was not found. This would ensure that this function always return at least one map. 
I understand why you removed it but I don't think that this input should be valid:
```
{ "path1": { "path2": [ {"field1": "value"},  {"field2": "value"}]}
```
Trying to insert `path1.path2.field1' or `path1.path2.field2' should throw an error imo because there are two maps that can contain these fields. The fact that they appear only once shouldn't matter, this layout is ambiguous.

Reply:
There are two separate points to discuss here:

- `extractSuffixMaps` should always return at least one map

I disagree with this one. It makes it much harder to act on the output in `insertValue`. As it is currently implemented, the logic in `insertValue` is simple and efficient. Emitting a map that we need to validate prior to modifying complicates the logic a lot. It's better to do all the validation in `extractSuffixMaps` and keep the modification logic in `insertValue` simple.

- `{ "path1": { "path2": [ {"field1": "value"},  {"field2": "value"}]}` should be considered ambiguous

What if someone wanted to add the path `path1.path2.field3`? Would that be considered ambiguous as well? If so, great, we can remain consistent in what is considered ambiguous and I see how to update the code to implement this ambiguity check. If not, we are inconsistent in what we consider ambiguous and I think that creates problems and confusion for users of this method.

Reply:
> I disagree with this one. It makes it much harder to act on the output in insertValue. As it is currently implemented, the logic in insertValue is simple and efficient. Emitting a map that we need to validate prior to modifying complicates the logic a lot. It's better to do all the validation in extractSuffixMaps and keep the modification logic in insertValue simple.

I am not sure I follow. I am just saying that if the path is not found we should return the parent map where the field should be added. That doesn't change the separation between `insertValue` and `extractSuffixMaps`. It feels wrong to me that calling `extractSuffixMaps` return an empty list for something like `extractSuffixMaps("field1", new HashMap<>())`.

> What if someone wanted to add the path path1.path2.field3? Would that be considered ambiguous as well?

Yes exactly, there are two possible maps for `path1.path2` so inserting any value there is ambiguous.

Reply:
Ok, I think I understand your approach now. Please confirm if I understand correctly:

Follow the path until no more path elements can be looked up in the source map. At that point, emit a suffix map where the suffix is either:

- The key of the value to replace in the map that was traversed to
- The path elements that could not be looked up

If we traverse to a list of maps, fail because it is ambiguous which map we should modify.

Putting it all together, if we had this source map:

```json
{
  "path1": {
    "path2.path3": {
      "key1": "value1"
  }
}
```

And we wanted to insert `"path1.path2.path3.key2": "value2"`, the result should be:

```json
{
  "path1": {
    "path2.path3": {
      "key1": "value1",
      "key2": "value2
  }
}
```

Do I understand correctly?

Reply:
> Do I understand correctly?

Yes thanks for clarifying.